katdal/caldata.py

```diff
@@ -0,0 +1,400 @@
+                                                                                x)
+        phase = phase_re + phase_im
+        return phase / np.abs(phase) * mag
+
+
+def interpolate_nans_1d(y):
+    """Interpolate over nans in a timeseries.
+
+    When interpolating the per-channel timeseries, the channels with more
+    nan values will deviate from the neighboring channels appearing as spikes
+    when the mean over time is plotted.
+    """
+    if np.isnan(y).all():
+        return y  # nothing to do, all nan values
+    if np.isfinite(y).all():
+        return y  # nothing to do, all values valid
+
+    # interpolate across nans
+    nan_locs = np.isnan(y)
+    xi = np.nonzero(~nan_locs)[0]
+    yi = y[xi]
+    x = np.arange(len(y))
+    y = _safe_linear_interp(xi, yi, x)
+    return y
+
+
+def _get_cal_sensor(key, katdal_obj):
+    """Load a calibrator solution from sensors.
+    """
+    values = []
+    timestamps = []
+    for sensor in katdal_obj.file['TelescopeState'].keys():
+        if key in sensor:
+            value = katdal_obj.sensor['TelescopeState/{}'.format(sensor)]
+            timestamp = katdal_obj.sensor.timestamps
+            try:
+                values = np.vstack((values, value))
+                timestamps = np.vstack((timestamps, timestamp))
+            except ValueError:
+                values = np.asarray(value)
+                timestamps = np.asarray(timestamp)
+    timestamps = np.reshape(timestamps, timestamps.size)
+    if timestamps.size > 1:
+        timestamps = timestamps.squeeze()
+    return timestamps, values
+
+
+def _get_cal_antlist(katdal_obj):
+    """Load antenna list used for calibration.
+    If the value does not match the antenna list in the katdal dataset,
+    a :exc:`CalibrationReadError` is raised. Eventually this could be
+    extended to allow for an antenna list that doesn't match by permuting
+    the calibration solutions.
+    """
+    cal_antlist = katdal_obj._get_telstate_attr('cal_antlist')
+    if isinstance(cal_antlist, np.ndarray):
+        cal_antlist = cal_antlist.tolist()
+    if cal_antlist != [ant.name for ant in katdal_obj.ants]:
+        raise CalibrationReadError('cal_antlist does not match katdal antenna list')
+    return cal_antlist
+
+
+def _get_cal_pol_ordering(katdal_obj):
+    """Load polarization ordering used by calibration solutions.
+
+    Returns
+    -------
+    dict
+        Keys are 'h' and 'v' and values are 0 and 1, in some order
+    """
+    cal_pol_ordering = katdal_obj._get_telstate_attr('cal_pol_ordering')
+    try:
+        cal_pol_ordering = np.array(cal_pol_ordering)
+    except (NameError, SyntaxError):
+        raise
+    except Exception as e:
+        raise CalibrationReadError(str(e))
+
+    if cal_pol_ordering.shape != (4, 2):
+        # assume newer telstate format
+        pol_dict = {}
+        for idx, pol in enumerate(cal_pol_ordering):
+            pol_dict[pol] = idx
```

Comment:
This could be written as a dict comprehension: `{pol: idx for idx, pol in enumerate(cal_pol_ordering)}`.

It could also be shared by the other path instead of `{order[0]: 0, order[1]: 1}`.